src/datachain/lib/signal_schema.py

```diff
@@ -117,40 +151,116 @@ def from_column_types(col_types: dict[str, Any]) -> "SignalSchema":
+        based on whether the type is Optional or not."""
+        orig = get_origin(fr_type)
+        args = get_args(fr_type)
+        # Check if fr_type is Optional
+        if orig == Union and len(args) == 2 and (type(None) in args):
+            fr_type = args[0]
+            orig = get_origin(fr_type)
+        if orig in (Literal, LiteralEx):
+            # Literal has no __name__ in Python 3.9
+            type_name = "Literal"
+        else:
+            type_name = str(fr_type.__name__)  # type: ignore[union-attr]
+        return type_name, fr_type
+
+    @staticmethod
+    def serialize_custom_model_fields(
+        name: str, fr: type, custom_types: dict[str, Any]
+    ) -> str:
+        """This serializes any custom type information to the provided custom_types
+        dict, and returns the name of the type provided."""
+        if (
+            hasattr(fr, "__origin__")
+            or not issubclass(fr, BaseModel)
+            or name in ("File", "TextFile", "ImageFile")
+        ):
+            # Don't store non-feature types, as well as built-in types.
+            return name
+        version_name = ModelStore.get_name(fr)
+        if version_name in custom_types:
+            # This type is already stored in custom_types.
+            return version_name
+        fields = {}
+        for field_name, info in fr.model_fields.items():
+            field_type = info.annotation
+            # All fields should be typed.
+            assert field_type
+            field_type_name, field_type = SignalSchema._get_name_original_type(
+                field_type
+            )
+            # Serialize this type to custom_types if it is a custom type as well.
+            fields[field_name] = SignalSchema.serialize_custom_model_fields(
+                field_type_name, field_type, custom_types
+            )
+        custom_types[version_name] = fields
+        return version_name
+
+    def serialize(self) -> dict[str, Any]:
+        signals: dict[str, Any] = {}
+        custom_types: dict[str, Any] = {}
         for name, fr_type in self.values.items():
             if (fr := ModelStore.to_pydantic(fr_type)) is not None:
                 ModelStore.register(fr)
                 signals[name] = ModelStore.get_name(fr)
+                type_name, fr_type = SignalSchema._get_name_original_type(fr)
             else:
-                orig = get_origin(fr_type)
-                args = get_args(fr_type)
-                # Check if fr_type is Optional
-                if orig == Union and len(args) == 2 and (type(None) in args):
-                    fr_type = args[0]
-                signals[name] = str(fr_type.__name__)  # type: ignore[union-attr]
+                type_name, fr_type = SignalSchema._get_name_original_type(fr_type)
+                signals[name] = type_name
+            self.serialize_custom_model_fields(type_name, fr_type, custom_types)
+        if custom_types:
```

Comment:
Am I missing something, or is `custom_types` always empty here?

Reply:
Should it be an attribute of `SignalSchema`? It looks like a local variable but doesn't seem to be used like one AFAICT.

Reply:
The `custom_types` dict is set within `serialize_custom_model_fields`, see above for example: https://github.com/iterative/datachain/pull/348/files/2a86ad17b910d39e197ea24e3fd4a507b62249d0#diff-69a1fc2d77ff106392ad7649ed8cf7ec344dad230af2e4d0ea9c64b816cee15bR200

I didn't want to just make this a return-based set as it would lead to extra code complexity (due to the recursive nature of `serialize_custom_model_fields`) and I didn't want to make this an attribute of `SignalSchema` as this information is only valid during the call to `serialize` and may become out of date or incorrect at any other times (and is not used at any other times).